src/dune_engine/process.ml

```diff
@@ -420,38 +456,33 @@ end = struct
       in
       Has_output { with_color; without_color; has_embedded_location }
 
-  (* In this module, we don't need the "Error: " prefix given that it is already
-     included in the error message from the command. *)
-  let fail ~output ~purpose ~dir paragraphs =
-    let paragraphs : User_message.Style.t Pp.t list =
-      match output with
-      | No_output -> paragraphs
-      | Has_output output -> paragraphs @ [ output.with_color ]
-    in
-    let dir =
-      match dir with
-      | None -> Path.of_string (Sys.getcwd ())
-      | Some dir -> dir
-    in
+  let get_loc_and_annots ~dir ~purpose ~output =
     let loc, annots = loc_and_annots_of_purpose purpose in
+    let dir = Option.value dir ~default:Path.root in
     let annots = With_directory_annot.make dir :: annots in
     let annots =
-      if has_embedded_location output then
-        let annots = User_error.Annot.Has_embedded_location.make () :: annots in
-        match
-          match output with
-          | No_output -> None
-          | Has_output output ->
-            Compound_user_error.parse_output ~dir output.without_color
-        with
-        | None -> annots
-        | Some annot -> annot :: annots
-      else
-        annots
+      match output with
+      | No_output -> annots
+      | Has_output output ->
+        if output.has_embedded_location then
+          let annots =
+            User_error.Annot.Has_embedded_location.make () :: annots
+          in
+          match Compound_user_error.parse_output ~dir output.without_color with
+          | None -> annots
+          | Some annot -> annot :: annots
+        else
+          annots
     in
+    (loc, annots)
+
+  let fail ~loc ~annots paragraphs =
+    (* We don't use [User_error.make] as it would add the "Error: " prefix. We
+       don't need this prefix as it is already included in the output of the
+       command. *)
```

Comment:
At one point I was wondering if we should move annotations to the `User_message` module so that below you would write `User_message.make ?loc paragraphs ~annots`. I find it a bit strange that we attach annotations to exceptions rather than to error messages. 

(This is not immediately relevant to this PR, just something I was wondering about.)

Reply:
I never thought about it, but yes I agree with you.